src/Faker/Provider/zh_CN/Text.php

```diff
@@ -0,0 +1,1063 @@
+    }
+
+    protected static function appendEnd($text)
+    {
+        $text = static::fixPairPunctuation($text);
+
+        $words = static::explode($text);
+
+        // remove all not end punctuations
+        $preventLoopTooMuchTimes = 0;
+        while (in_array($words[count($words) - 1], static::$notEndPunctuations) && $preventLoopTooMuchTimes < 10) {
+            array_pop($words);
+            $preventLoopTooMuchTimes++;
+        }
+
+        // if the last char is not a valid punctuation, append a default one
+        if (!in_array($words[count($words) - 1], static::$punctuations)) {
+            $words[] = static::$defaultEndPunctuations;
+        }
+
+        return static::implode($words);
+    }
+
+    protected static function fixPairPunctuation($text)
+    {
+        $words = static::explode($text);
+
+        $expectedPunctuation = array();
+        $expectedPunctuationInsertPos = array();
+        $insertPunctuation = array();
+        $lastInsertPos = 0;
+        $lastCharIsPunctuation = '';
+
+        foreach ($words as $pos => $word) {
+            if (array_key_exists($word, static::$punctuationPairs)) {
+                $expectedPunctuation[] = static::$punctuationPairs[$word];
+                $expectedPunctuationInsertPos[] = -1;
+            } elseif (in_array($word, static::$punctuationPairs)) {
+                if (!empty($expectedPunctuation) && $word === $expectedPunctuation[count($expectedPunctuation) - 1]) {
+                    array_pop($expectedPunctuation);
+                    array_pop($expectedPunctuationInsertPos);
+                } else {
+                    $insertPunctuation[$lastInsertPos] = array_search($word, static::$punctuationPairs);
+                }
+            } elseif (in_array($word, static::$punctuations)) {
+                if (!$lastCharIsPunctuation) {
+                    if (!empty($expectedPunctuationInsertPos) && $expectedPunctuationInsertPos[count($expectedPunctuationInsertPos) - 1] < 0) {
+                        $expectedPunctuationInsertPos[count($expectedPunctuationInsertPos) - 1] = $pos;
+                    }
+                }
+                $lastCharIsPunctuation = true;
+            } else {
+                if ($lastCharIsPunctuation) {
+                    $lastInsertPos = $pos;
+                }
+                $lastCharIsPunctuation = false;
+            }
+        }
+
+        foreach ($expectedPunctuationInsertPos as $index => $pos) {
+            $insertPunctuation[$pos] = $expectedPunctuation[$index];
+        }
+
+        foreach ($insertPunctuation as $pos => $punctuation) {
+            $words[$pos] = $punctuation . $words[$pos];
+        }
+
+        return static::implode($words);
+    }
+
+    /**
+     * UTF-8 strlen
+     *
+     * @see https://en.wikipedia.org/wiki/UTF-8#Description
+     *
+     * @param string $text
+     *
+     * @return int
+     */
+    protected static function utf8Strlen($text)
```

Comment:
wont the php native mb functions handle this properly nowadays?

Reply:
I'm happy to change all this messy code to `mb_*` function. But there is no `ext_mbstring` in package.json. So I don't use these functions.